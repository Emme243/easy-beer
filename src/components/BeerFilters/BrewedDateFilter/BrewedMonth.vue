<template>
  <v-dialog
    v-model="isModalOpen"
    overlay-opacity="0.3"
    :close-on-content-click="false"
    transition="scale-transition"
    offset-y
    max-width="290px"
    min-width="auto"
  >
    <template #activator="{ on, attrs }">
      <v-text-field
        :value="brewedMonth"
        :label="inputLabel"
        readonly
        :append-icon="brewedMonth && 'fa fa-arrow-rotate-left'"
        v-bind="attrs"
        v-on="on"
        @click:append="handleBrewedMonthChange('')"
        data-test-id="brewed-month-text-field"
      ></v-text-field>
    </template>
    <v-date-picker
      type="month"
      scrollable
      @input="handleBrewedMonthChange"
      :value="brewedMonth"
      :min="brewedMinMonth"
      :max="brewedMaxMonth"
      show-current="2014-01"
      data-test-id="brewed-month-date-picker"
    >
      <v-spacer></v-spacer>
      <v-btn
        text
        color="primary"
        @click="handleBrewedMonthChange('')"
        data-test-id="reset-brewed-month-btn"
      >
        Limpiar
      </v-btn>
      <v-btn
        text
        color="primary"
        @click="isModalOpen = false"
        data-test-id="close-date-picker-button"
      >
        OK
      </v-btn>
    </v-date-picker>
  </v-dialog>
</template>

<script>
import { mapActions } from 'vuex';

export default {
  name: 'BrewedMonth',
  props: {
    inputLabel: {
      type: String,
      required: true,
    },
    brewedMonth: {
      type: String,
      required: true,
    },
    brewedMinMonth: {
      type: String,
      default: '',
    },
    brewedMaxMonth: {
      type: String,
      default: '',
    },
  },
  data: () => ({ isModalOpen: false }),
  methods: {
    ...mapActions({ resetPageInStore: 'filter/resetPage' }),
    handleBrewedMonthChange(month) {
      this.$emit('setBrewedMonthInStore', month);
      this.resetPageInStore();
    },
  },
};
</script>

<style scoped></style>
