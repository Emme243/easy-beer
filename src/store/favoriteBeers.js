export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
