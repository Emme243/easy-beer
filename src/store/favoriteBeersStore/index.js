export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
