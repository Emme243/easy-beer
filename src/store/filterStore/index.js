import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
