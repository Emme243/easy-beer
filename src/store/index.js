import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
