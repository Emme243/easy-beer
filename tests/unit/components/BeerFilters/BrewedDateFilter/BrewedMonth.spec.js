import BrewedMonth from '@/components/BeerFilters/BrewedDateFilter/BrewedMonth';
import store from '@/store';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';

describe('Pruebas en el componente BrewedMonth.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  const minBrewedMonthProps = {
    brewedMonth: '02-2014',
    inputLabel: 'Mes de elaboración de la cerveza',
    brewedMaxMonth: '10-2014',
  };

  const mountFunction = options => {
    return shallowMount(BrewedMonth, {
      localVue,
      vuetify,
      store,
      propsData: minBrewedMonthProps,
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const resetButtonSelector = '[data-test-id="reset-brewed-month-btn"]';
  const monthPickerSelector = '[data-test-id="brewed-month-date-picker"]';
  const closeModalButtonSelector = '[data-test-id="close-date-picker-button"]';
  const brewedMonthTextField = '[data-test-id="brewed-month-text-field"]';

  it('Debe hacer match con el snapshot', function () {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('El botón de limpiar debe emitir el evento setBrewedMonthInStore con un string vacío y llamar al dispatch resetPage cuando se hace click', async function () {
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const resetButton = wrapper.find(resetButtonSelector);
    expect(resetButton.exists()).toBeTruthy();
    resetButton.vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('setBrewedMonthInStore').length).toBe(1);
    expect(wrapper.emitted('setBrewedMonthInStore')[0][0]).toBe('');
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetPage');
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('El datepicker debe emitir el evento setBrewedMonthInStore con el nuevo mes y llamar al dispatch resetPage cuando se hace click', async function () {
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const monthPicker = wrapper.find(monthPickerSelector);
    expect(monthPicker.exists()).toBeTruthy();
    monthPicker.vm.$emit('input', '02-2014');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('setBrewedMonthInStore').length).toBe(1);
    expect(wrapper.emitted('setBrewedMonthInStore')[0][0]).toBe('02-2014');
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetPage');
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('El botón de OK debe cerrar el datepicker cuando se hace click, la propiedad isModalOpen debe ser false', async function () {
    const wrapper = mountFunction({ data: () => ({ isModalOpen: true }) });
    expect(wrapper.vm.isModalOpen).toBeTruthy();
    const closeModalButton = wrapper.find(closeModalButtonSelector);
    expect(closeModalButton.exists()).toBeTruthy();
    closeModalButton.vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isModalOpen).toBeFalsy();
  });
});
