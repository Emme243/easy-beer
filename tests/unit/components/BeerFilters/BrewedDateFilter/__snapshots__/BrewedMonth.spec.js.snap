// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente BrewedMonth.vue Debe hacer match con el snapshot 1`] = `
<v-dialog-stub contentclass="" overlayopacity="0.3" opendelay="0" closedelay="0" openonclick="true" maxwidth="290px" origin="center center" retainfocus="true" transition="scale-transition" offset-y="" min-width="auto">
  <v-date-picker-stub width="290" eventcolor="warning" firstdayofweek="0" localefirstdayofyear="0" max="10-2014" min="" nexticon="$next" nextmontharialabel="$vuetify.datePicker.nextMonthAriaLabel" nextyeararialabel="$vuetify.datePicker.nextYearAriaLabel" previcon="$prev" prevmontharialabel="$vuetify.datePicker.prevMonthAriaLabel" prevyeararialabel="$vuetify.datePicker.prevYearAriaLabel" scrollable="true" showcurrent="2014-01" selecteditemstext="$vuetify.datePicker.itemsSelected" type="month" value="02-2014" data-test-id="brewed-month-date-picker">
    <v-spacer-stub tag="div"></v-spacer-stub>
    <v-btn-stub color="primary" tag="button" activeclass="" text="true" type="button" data-test-id="reset-brewed-month-btn">
      Limpiar
    </v-btn-stub>
    <v-btn-stub color="primary" tag="button" activeclass="" text="true" type="button" data-test-id="close-date-picker-button">
      OK
    </v-btn-stub>
  </v-date-picker-stub>
</v-dialog-stub>
`;
